[lldb/docs] Add scripting extensions documentation to the website

medismailben opened the pull request:
This patch adds the documentation for a subset of scripting extensions such as scripted process, scripted thread, operating system threads & scritped thread plans to the lldb website.

JDevlieghere commented on lldb/docs/use/python.rst:
```diff
@@ -11,6 +11,65 @@ some of these things by going through an example, explaining how to use
 Python scripting to find a bug in a program that searches for text in a
 large binary tree.
 
+Operating System Thread Plugins
+-------------------------------
+
+.. automodapi:: lldb.plugins.operating_system
+    :no-heading:
+    :members:
+    :undoc-members:
+    :skip: command
+    :skip: declaration
+    :skip: in_range
+    :skip: is_numeric_type
+    :skip: lldb_iter
+    :no-inheritance-diagram:
+
+Scripted Process Plugins
+-------------------------------
+
+.. automodapi:: lldb.plugins.scripted_process
+    :no-heading:
+    :members:
+    :undoc-members:
+    :skip: ABCMeta
+    :skip: command
+    :skip: declaration
+    :skip: in_range
+    :skip: is_numeric_type
+    :skip: lldb_iter
+    :no-inheritance-diagram:
```
I don't think you meant to add this to the [Python Scripting](https://lldb.llvm.org/use/python.html) page as this is a tutorial. The automatically generated documentation should probably end up on the [Python API](https://lldb.llvm.org/python_api.html) page. But you probably also want to have a page that explains what these things are, and how you'd use them. It seems like the [Python Reference](https://lldb.llvm.org/use/python-reference.html) page would be best suited for that.

jimingham replied:
We really should refresh the Python Scripting page, having more task-focused docs would be good.  But that's pretty old at this point, and focused on a single problem, so I agree with Jonas, it's not the right place to put this info.

JDevlieghere reviewed (commented):
I'm excited to see these extension points getting documented! What would make this even more valuable is having a page that documents all of LLDB's extension points (and explains the duality between scripting LLDB interactions vs scripting LLDB's behavior) and explain what you can use them for and how to do that. Basically the points you made in your EuroLLVM talk, but in website form. The Python Reference page already covers a bit of that so that could serve as a starting point.

jimingham commented on lldb/examples/python/templates/scripted_thread_plan.py:
```diff
@@ -0,0 +1,70 @@
+        self.thread_plan = thread_plan
+
+    def explains_stop(self, event: lldb.SBEvent) -> bool:
+        """Each plan is asked if it "explains" the stop.  The first plan to claim the stop wins.
```
Might be good to specify the ordering here:

Each plan is asked - from youngest to oldest - if it "explains" the stop.

jimingham commented on lldb/examples/python/templates/scripted_thread_plan.py:
```diff
@@ -0,0 +1,70 @@
+            bool: `True` if this thread plan is stale, `False` otherwise.
+            Defaults to `True`.
+        """
+        return True
```
Should the template default to True here?  Plans don't frequently go stale, the get popped because of MischiefManaged much more frequently.  So most plans don't need to implement is_stale, at least not at first.

medismailben replied:
I went with `True` since this is the default value in case of an error on the C++ side:

https://github.com/apple/llvm-project/blob/swift/release/6.0/lldb/source/Plugins/ScriptInterpreter/Python/ScriptInterpreterPython.cpp#L1754

Do you think we should change the default value everywhere to be False ?

jimingham replied:
Saying you are stale in case of error is appropriate for the Scripted Thread Plan runner plan.  If the underlying Python code can't figure out what its state is, it's likely in a bad way and we should mark it as stale and get rid of it.  
But these templates are more analogous to the base ThreadPlan than to the scripted plan runner.

That one does:

  virtual bool IsPlanStale() { return false; }

bulbazord commented on lldb/docs/CMakeLists.txt:
```diff
@@ -25,10 +25,17 @@ if (LLDB_ENABLE_PYTHON AND SPHINX_FOUND)
     # Pretend that the SWIG generated API is a Python package.
     file(MAKE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/lldb)
     get_target_property(lldb_bindings_dir swig_wrapper_python BINARY_DIR)
+
```
nit: stray space?

bulbazord commented on lldb/docs/python_extensions.rst:
```diff
@@ -0,0 +1,36 @@
+=================
+
+LLDB provides many scriptable extensions to augment the debugger capabilities
+and give the ability to the user to tailor their experience to their own needs.
```
Suggestion: `LLDB provides scriptable extensions to augment the debugger's capabilities. This gives users the ability to tailor their debugging experience to their own needs.`

bulbazord commented on lldb/docs/python_extensions.rst:
```diff
@@ -0,0 +1,36 @@
+LLDB provides many scriptable extensions to augment the debugger capabilities
+and give the ability to the user to tailor their experience to their own needs.
+
+This page describes some of these scripting extension:
```
`extension` -> `extensions`

bulbazord commented on lldb/docs/python_extensions.rst:
```diff
@@ -0,0 +1,36 @@
+
+Operating System Thread Plugins
+-------------------------------
+
```
A short blurb explaining what each of these are for would be useful. It may not be obvious what "Operating System Thread Plugins" are for, as an example.

medismailben replied:
![image](https://github.com/llvm/llvm-project/assets/10001166/0dec76ec-ae53-4ea1-b5f5-b96e64cfb7de)

We get the blurb from the docstrings.

bulbazord replied:
Ah, this looks great! Thanks for showing me what it will look like :)

bulbazord commented on lldb/examples/python/templates/scripted_platform.py:
```diff
@@ -10,9 +10,6 @@ class ScriptedPlatform(metaclass=ABCMeta):
     overwritten by the inheriting class.
-
-    DISCLAIMER: THIS INTERFACE IS STILL UNDER DEVELOPMENT AND NOT STABLE.
-                THE METHODS EXPOSED MIGHT CHANGE IN THE FUTURE.
```
Why remove the disclaimer? Is this no longer true?

medismailben replied:
I'd expect at this point the extension to be stable.

bulbazord commented on lldb/examples/python/templates/scripted_thread_plan.py:
```diff
@@ -0,0 +1,70 @@
+from abc import abstractmethod
```
This is an entire scripting interface. Maybe a separate PR would be more appropriate? Seems like you're doing multiple things here.

bulbazord reviewed (commented):
I know you've been thinking about and working on this for a while, so glad to see it come together like this! 😃

medismailben committed:
[lldb/docs] Add scripting extensions documentation to the website

This patch adds the documentation for a subset of scripting extensions
such as scripted process, scripted thread, operating system threads &
scritped thread plans to the lldb website.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

bulbazord reviewed (approved):
LGTM
